scwx-qt/source/scwx/qt/manager/marker_manager.cpp

```diff
@@ -358,13 +430,56 @@ void MarkerManager::move_marker(size_t from, size_t to)
 
 void MarkerManager::for_each(std::function<MarkerForEachFunc> func)
 {
-   std::shared_lock lock(p->markerRecordLock_);
+   const std::shared_lock lock(p->markerRecordLock_);
    for (auto marker : p->markerRecords_)
    {
       func(marker->markerInfo_);
    }
 }
 
+void MarkerManager::add_icon(const std::string& name, bool startup)
+{
+   {
+      const std::unique_lock lock(p->markerIconsLock_);
+      if (p->markerIcons_.contains(name))
+      {
+         return;
+      }
+      const std::shared_ptr<boost::gil::rgba8_image_t> image =
+         ResourceManager::LoadImageResource(name);
+
+      auto icon = types::MarkerIconInfo(name, -1, -1, image);
+      p->markerIcons_.emplace(name, icon);
+   }
+
+   if (!startup)
+   {
+      util::TextureAtlas& textureAtlas = util::TextureAtlas::Instance();
+      textureAtlas.BuildAtlas(
+         2048, 2048); // Should this code be moved to ResourceManager?
+      Q_EMIT IconAdded(name);
+   }
+}
+
+std::optional<types::MarkerIconInfo>
+MarkerManager::get_icon(const std::string& name)
+{
+   const std::shared_lock lock(p->markerIconsLock_);
+   if (p->markerIcons_.contains(name))
```

Comment:
Using `find()`, and dereferencing the iterator if valid can be more efficient, since it only searches the map once.

```C++
   auto it = p->markerIcons_.find(name);
   if (it != p->markerIcons_.end())
   {
      return it->second;
   }
```